File: VIQA/HtmlElements/Interfaces/ITable.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using VIQA.Common;
using VIQA.HtmlElements.SimpleElements;

namespace VIQA.HtmlElements.Interfaces
{
    public interface ITable<T> : IHaveValue where T : VIElementsSet, IHaveValue
    {
        Cell<T> Cell(int colNum, int rowNum);
        Cell<T> Cell(string colName, int rowNum);
        Cell<T> Cell(int colNum, string rowName);
        Cell<T> Cell(string colName, string rowName);
        
        List<Cell<T>> FindCellsWithValue(string value);
        List<Cell<T>> FindCellsWithValue(Regex regex);
        Cell<T> FindFirstCellWithValue(string value);
        Cell<T> FindCellInColumn(int colIndex, string value);
        List<Cell<T>> FindCellsInColumn(int colIndex, Regex regex);
        Cell<T> FindCellInColumn(string colName, string value);
        List<Cell<T>> FindCellsInColumn(string colname, Regex regex);
        Cell<T> FindCellInRow(int rowIndex, string value);
        List<Cell<T>> FindCellsInRow(int rowIndex, Regex regex);
        Cell<T> FindCellInRow(string rowName, string value);
        List<Cell<T>> FindCellsInRow(string rowName, Regex regex);
        List<Cell<T>> FindColumnByRowValue(int rowIndex, string value);
        List<Cell<T>> FindColumnByRowValue(string rowName, string value);
        List<Cell<T>> FindRowByColumnValue(int colIndex, string value);
        List<Cell<T>> FindRowByColumnValue(string colName, string value);

        Columns<T> Columns { get; }
        Rows<T> Rows { get; }
        string[] Footer { get; set; }
    }

    public interface ITable : ITable<TextElement> { }
}

File: Java/src/main/java/ru/viqa/ui_testing/elements/interfaces/ITable.java
package ru.viqa.ui_testing.elements.interfaces;

import ru.viqa.ui_testing.elements.baseClasses.Table.Cell;
import ru.viqa.ui_testing.elements.baseClasses.Table.Columns;
import ru.viqa.ui_testing.elements.baseClasses.Table.Rows;

import java.util.List;

/**
 * Created by roman.i on 20.10.2014.
 */

public interface ITable<T extends IHaveValue> extends IHaveValue {
    Cell<T> cell(int colNum, int rowNum) throws Exception;
    Cell<T> cell(String colName, int rowNum) throws Exception;
    Cell<T> cell(int colNum, String rowName) throws Exception;
    Cell<T> cell(String colName, String rowName) throws Exception;
    List<Cell<T>> findCellsValues(String value) throws Exception;
    List<Cell<T>> matchCellsValues(String regex) throws Exception;
    Cell<T> findFirstCellWithValue(String value) throws Exception;

    Cell<T> findCellInColumn(int colIndex, String value) throws Exception;
    List<Cell<T>> matchCellsInColumn(int colIndex, String regex) throws Exception;
    Cell<T> findCellInColumn(String colName, String value) throws Exception;
    List<Cell<T>> matchCellsInColumn(String colname, String regex) throws Exception;

    Cell<T> findCellInRow(int rowIndex, String value) throws Exception;
    List<Cell<T>> matchCellsInRow(int rowIndex, String regex) throws Exception;
    Cell<T> findCellInRow(String rowName, String value) throws Exception;
    List<Cell<T>> matchCellsInRow(String rowName, String pattern) throws Exception;

    List<Cell<T>> findColumnByRowValue(int rowIndex, String value) throws Exception;
    List<Cell<T>> findColumnByRowValue(String rowName, String value) throws Exception;
    List<Cell<T>> findRowByColumnValue(int colIndex, String value) throws Exception;
    List<Cell<T>> findRowByColumnValue(String colName, String value) throws Exception;

    Columns<T> getColumns();
    List<Cell<T>> getColumn(int colNum) throws Exception;
    List<Cell<T>> getColumn(String colName) throws Exception;
    Rows<T> getRows();
    List<Cell<T>> getRow(int rowNum) throws Exception;
    List<Cell<T>> getRow(String rowName) throws Exception;
    String[] getFooter() throws Exception;
}
